File: app/src/mock/java/com/amandariu/tagger/data/FakeTagsRemoteDataSource.java
package com.amandariu.tagger.data;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import com.amandariu.tagger.demo.data.Tag;
import com.amandariu.tagger.demo.data.source.TagsDataSource;
import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.List;

/**
 * Implementation of a remote datasource for easy testing.
 *
 * @author Amanda Riu
 */
public class FakeTagsRemoteDataSource implements TagsDataSource {

    private static FakeTagsRemoteDataSource sInstance = null;
    private static final List<Tag> TAGS_DATA_SERVICE = new ArrayList<>();


    public static FakeTagsRemoteDataSource getInstance() {
        if (sInstance == null) {
            sInstance = new FakeTagsRemoteDataSource();
        }
        return sInstance;
    }

    /**
     * Load tags from one or more datasources.
     *
     * @param callback The callback to send the loaded tags to.
     */
    @Override
    public void getTags(@NonNull LoadTagsCallback callback) {
        callback.onTagsLoaded(Lists.newArrayList(TAGS_DATA_SERVICE));
    }

    /**
     *
     */
    @Override
    public void refreshTags() {
        //
        // Not needed.
    }

    @VisibleForTesting
    public void addTags(Tag... tags) {
        if (tags != null) {
            for (Tag tag : tags) {
                TAGS_DATA_SERVICE.add(tag);
            }
        }
    }
}

File: app/src/mock/java/com/amandariu/tagger/data/FakeTagsRemoteDataSource.kt
package com.amandariu.tagger.data

import android.support.annotation.VisibleForTesting
import android.util.Log
import com.amandariu.tagger.ITag
import com.amandariu.tagger.demo.data.Tag
import com.amandariu.tagger.demo.data.source.ISourceBase
import com.amandariu.tagger.demo.data.source.ITagsDataSource
import com.bluelinelabs.logansquare.LoganSquare
import com.google.common.collect.Lists
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

class FakeTagsRemoteDataSource : ITagsDataSource {
    companion object {
        private var sInstance: ITagsDataSource? = null
        private val TAGS_DATA_SERVICE = arrayListOf<Tag>()

        @JvmStatic
        fun getInstance(): ITagsDataSource {
            if (sInstance == null) {
                sInstance = FakeTagsRemoteDataSource()
            }
            return sInstance!!
        }
    }
    init {
        addTags()
    }

    override fun saveTags(tags: List<ITag>, callback: ITagsDataSource.ISaveTagsCallback) {
        // todo
    }

    override fun getTags(callback: ISourceBase.ILoadTagsCallback) {
        callback.onTagsLoaded(Lists.newArrayList(TAGS_DATA_SERVICE))
    }

    override fun shutdown() {
        // todo
    }

    @VisibleForTesting
    private fun addTags() {
        val tagJson = getStringFromResourceFile("sample_tags.json")
        val tags = LoganSquare.parseList(tagJson, Tag::class.java)
        for (tag in tags) {
            TAGS_DATA_SERVICE.add(tag)
        }
    }

    private fun getStringFromResourceFile(filename: String): String? {
        try {
            val inStream = this.javaClass.classLoader.getResourceAsStream(filename)
            val bufferedReader = BufferedReader(InputStreamReader(inStream, "UTF-8"))

            val buffer = StringBuilder()
            var lineString: String? = null

            while ({ lineString = bufferedReader.readLine(); lineString }() != null) {
                buffer.append(lineString)
            }

            bufferedReader.close()
            return buffer.toString()
        } catch (e: IOException) {
            Log.d("AMANDA-TEST", "Could not load contents of json file")
            return null
        }
    }
}
